Null Pointer in org.graylog2.periodical.IndexerClusterCheckerThread

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->
No null Pointer
## Current Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->

```
2022-06-09 10:10:51,556 ERROR: org.graylog2.periodical.IndexerClusterCheckerThread - Uncaught exception in periodical
org.graylog.shaded.elasticsearch7.org.elasticsearch.ElasticsearchException: Unable to retrieve nodes list
        at org.graylog.storage.elasticsearch7.ElasticsearchClient.exceptionFrom(ElasticsearchClient.java:140) ~[?:?]
        at org.graylog.storage.elasticsearch7.ElasticsearchClient.execute(ElasticsearchClient.java:100) ~[?:?]
        at org.graylog.storage.elasticsearch7.cat.CatApi.perform(CatApi.java:89) ~[?:?]
        at org.graylog.storage.elasticsearch7.cat.CatApi.nodes(CatApi.java:51) ~[?:?]
        at org.graylog.storage.elasticsearch7.ClusterAdapterES7.nodes(ClusterAdapterES7.java:107) ~[?:?]
        at org.graylog.storage.elasticsearch7.ClusterAdapterES7.fileDescriptorStats(ClusterAdapterES7.java:100) ~[?:?]
        at org.graylog2.indexer.cluster.Cluster.getFileDescriptorStats(Cluster.java:90) ~[graylog.jar:?]
        at org.graylog2.periodical.IndexerClusterCheckerThread.checkOpenFiles(IndexerClusterCheckerThread.java:75) ~[graylog.jar:?]
        at org.graylog2.periodical.IndexerClusterCheckerThread.doRun(IndexerClusterCheckerThread.java:64) ~[graylog.jar:?]
        at org.graylog2.plugin.periodical.Periodical.run(Periodical.java:77) [graylog.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.graylog.storage.elasticsearch7.cat.NodeResponse`, problem: Null diskUsed
 at [Source: (org.graylog.shaded.elasticsearch7.org.apache.http.nio.entity.ContentInputStream); line: 1, column: 1615] (through reference chain: java.util.ArrayList[8])
        at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1608) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:484) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:503) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:285) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromObjectWith(ValueInstantiator.java:229) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:195) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:488) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:286) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4014) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3078) ~[graylog.jar:?]
        at org.graylog.storage.elasticsearch7.cat.CatApi.returnType(CatApi.java:98) ~[?:?]
        at org.graylog.storage.elasticsearch7.cat.CatApi.lambda$perform$4(CatApi.java:93) ~[?:?]
        at org.graylog.storage.elasticsearch7.ElasticsearchClient.execute(ElasticsearchClient.java:98) ~[?:?]
        ... 14 more
Caused by: java.lang.NullPointerException: Null diskUsed
        at org.graylog.storage.elasticsearch7.cat.AutoValue_NodeResponse.<init>(AutoValue_NodeResponse.java:55) ~[?:?]
        at org.graylog.storage.elasticsearch7.cat.NodeResponse.create(NodeResponse.java:58) ~[?:?]
        at jdk.internal.reflect.GeneratedMethodAccessor71.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at com.fasterxml.jackson.databind.introspect.AnnotatedMethod.call(AnnotatedMethod.java:109) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:283) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromObjectWith(ValueInstantiator.java:229) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:195) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:488) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:286) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4014) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3078) ~[graylog.jar:?]
        at org.graylog.storage.elasticsearch7.cat.CatApi.returnType(CatApi.java:98) ~[?:?]
        at org.graylog.storage.elasticsearch7.cat.CatApi.lambda$perform$4(CatApi.java:93) ~[?:?]
        at org.graylog.storage.elasticsearch7.ElasticsearchClient.execute(ElasticsearchClient.java:98) ~[?:?]
        ... 14 more
```
## Possible Solution
<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
1. unclear how it can be reproduced
2.
3.
4.

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->

* Graylog Version: 4.2.7
* Java Version: JDK 11
* Elasticsearch Version: 7.10.2
* MongoDB Version:
* Operating System:
* Browser version:

--- a/graylog-storage-elasticsearch6/src/main/java/org/graylog/storage/elasticsearch6/ClusterAdapterES6.java
+++ b/graylog-storage-elasticsearch6/src/main/java/org/graylog/storage/elasticsearch6/ClusterAdapterES6.java
@@ -92,8 +92,21 @@ public class ClusterAdapterES6 implements ClusterAdapter {
                 final String host = jsonElement.path("host").asText(null);
                 final String ip = jsonElement.path("ip").asText();
                 final JsonNode fileDescriptorMax = jsonElement.path("fileDescriptorMax");
-                final Long maxFileDescriptors = fileDescriptorMax.isLong() ? fileDescriptorMax.asLong() : null;
-                setBuilder.add(NodeFileDescriptorStats.create(name, ip, host, maxFileDescriptors));
+                if (!fileDescriptorMax.isLong()) {
+                    try {
+                        //try parsing the String, it may represent a Long value
+                        setBuilder.add(NodeFileDescriptorStats.create(name, ip, host, Long.parseLong(fileDescriptorMax.asText())));
+                    } catch (Exception e) {
+                        StringBuilder sb = new StringBuilder("{");
+                        sb.append("name = ").append(name);
+                        sb.append(", host = ").append(host);
+                        sb.append(", ip = ").append(ip);
+                        sb.append("}");
+                        LOG.info("_cat/nodes API has returned a node without disk statistics: " + sb);
+                    }
+                } else {
+                    setBuilder.add(NodeFileDescriptorStats.create(name, ip, host, fileDescriptorMax.asLong()));
+                }
             }
         }
 
@@ -106,17 +119,29 @@ public class ClusterAdapterES6 implements ClusterAdapter {
         final ImmutableSet.Builder<NodeDiskUsageStats> setBuilder = ImmutableSet.builder();
         for (JsonNode jsonElement : nodes) {
             if (jsonElement.isObject()) {
-                setBuilder.add(
-                        NodeDiskUsageStats.create(
-                                jsonElement.path("name").asText(),
-                                jsonElement.path("role").asText(),
-                                jsonElement.path("ip").asText(),
-                                jsonElement.path("host").asText(null),
-                                jsonElement.path("diskUsed").asText(),
-                                jsonElement.path("diskTotal").asText(),
-                                jsonElement.path("diskUsedPercent").asDouble(NodeDiskUsageStats.DEFAULT_DISK_USED_PERCENT)
-                        )
-                );
+                final JsonNode diskUsed = jsonElement.path("diskUsed");
+                final JsonNode diskTotal = jsonElement.path("diskTotal");
+                if (!diskTotal.isMissingNode() && !diskUsed.isMissingNode()) {
+                    setBuilder.add(
+                            NodeDiskUsageStats.create(
+                                    jsonElement.path("name").asText(),
+                                    jsonElement.path("role").asText(),
+                                    jsonElement.path("ip").asText(),
+                                    jsonElement.path("host").asText(null),
+                                    diskUsed.asText(),
+                                    diskTotal.asText(),
+                                    jsonElement.path("diskUsedPercent").asDouble(NodeDiskUsageStats.DEFAULT_DISK_USED_PERCENT)
+                            )
+                    );
+                } else {
+                    StringBuilder sb = new StringBuilder("{");
+                    sb.append("name = ").append(jsonElement.path("name").asText());
+                    sb.append(", host = ").append(jsonElement.path("host").asText(null));
+                    sb.append(", ip = ").append(jsonElement.path("ip").asText());
+                    sb.append(", role = ").append(jsonElement.path("role").asText());
+                    sb.append("}");
+                    LOG.info("_cat/nodes API has returned a node without disk statistics: " + sb);
+                }
             }
         }
         return setBuilder.build();

--- a/graylog-storage-elasticsearch6/src/main/java/org/graylog/storage/elasticsearch6/ClusterAdapterES6.java
+++ b/graylog-storage-elasticsearch6/src/main/java/org/graylog/storage/elasticsearch6/ClusterAdapterES6.java
@@ -96,7 +96,7 @@ public class ClusterAdapterES6 implements ClusterAdapter {
                     try {
                         //try parsing the String, it may represent a Long value
                         setBuilder.add(NodeFileDescriptorStats.create(name, ip, host, Long.parseLong(fileDescriptorMax.asText())));
-                    } catch (Exception e) {
+                    } catch (NumberFormatException e) {
                         StringBuilder sb = new StringBuilder("{");
                         sb.append("name = ").append(name);
                         sb.append(", host = ").append(host);

--- a/graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/ClusterAdapterES6Test.java
+++ b/graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/ClusterAdapterES6Test.java
@@ -1,3 +1,19 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
 package org.graylog.storage.elasticsearch6;
 
 import com.fasterxml.jackson.databind.JsonNode;

--- a/graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/ClusterAdapterES6Test.java
+++ b/graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/ClusterAdapterES6Test.java
@@ -1,0 +1,105 @@
+package org.graylog.storage.elasticsearch6;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.joschi.jadconfig.util.Duration;
+import io.searchbox.client.JestClient;
+import io.searchbox.core.CatResult;
+import org.graylog2.indexer.cluster.health.NodeDiskUsageStats;
+import org.graylog2.indexer.cluster.health.NodeFileDescriptorStats;
+import org.graylog2.indexer.cluster.health.NodeRole;
+import org.graylog2.indexer.cluster.health.SIUnitParser;
+import org.graylog2.shared.bindings.providers.ObjectMapperProvider;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.Set;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+
+class ClusterAdapterES6Test {
+
+    private final static String SAMPLE_CAT_NODES_RESPONSE = "{" +
+            "\"result\" : " +
+            "[" +
+            "{" +
+            "\"id\":\"nodeWithCorrectInfo\"," +
+            "\"name\":\"nodeWithCorrectInfo\"," +
+            "\"role\":\"dimr\"," +
+            "\"ip\":\"182.88.0.2\"," +
+            "\"fileDescriptorMax\":1048576," +
+            "\"diskUsed\":\"45gb\"," +
+            "\"diskTotal\":\"411.5gb\"," +
+            "\"diskUsedPercent\":\"10.95\"" +
+            "}," +
+            "{" +
+            "\"id\":\"nodeWithMissingDiskStatistics\"," +
+            "\"name\":\"nodeWithMissingDiskStatistics\"," +
+            "\"role\":\"dimr\"," +
+            "\"ip\":\"182.88.0.1\"" +
+            "}" +
+            "]" +
+            "}";
+
+    private ClusterAdapterES6 clusterAdapter;
+
+    @BeforeEach
+    void setUp() throws Exception {
+        final JestClient jestClient = mock(JestClient.class);
+        final CatResult catResult = mock(CatResult.class);
+        final ObjectMapper objectMapper = new ObjectMapperProvider().get();
+        final JsonNode catApiJsonResponse = objectMapper.readTree(SAMPLE_CAT_NODES_RESPONSE);
+
+        doReturn(catResult).when(jestClient).execute(any());
+        doReturn(catApiJsonResponse).when(catResult).getJsonObject();
+        doReturn(true).when(catResult).isSucceeded();
+        
+        this.clusterAdapter = new ClusterAdapterES6(jestClient, Duration.minutes(1));
+    }
+
+    @Test
+    void testFileDescriptorStats() {
+        final Set<NodeFileDescriptorStats> nodeFileDescriptorStats = clusterAdapter.fileDescriptorStats();
+
+        assertThat(nodeFileDescriptorStats)
+                .hasSize(1)
+                .noneSatisfy(
+                        nodeDescr -> assertThat(nodeDescr.name()).isEqualTo("nodeWithMissingDiskStatistics")
+                )
+                .first()
+                .satisfies(
+                        nodeDescr -> {
+                            assertThat(nodeDescr.name()).isEqualTo("nodeWithCorrectInfo");
+                            assertThat(nodeDescr.ip()).isEqualTo("182.88.0.2");
+                            assertThat(nodeDescr.fileDescriptorMax()).isPresent();
+                            assertThat(nodeDescr.fileDescriptorMax().get()).isEqualTo(1048576L);
+                        }
+                );
+    }
+
+    @Test
+    void testDiskUsageStats() {
+        final Set<NodeDiskUsageStats> diskUsageStats = clusterAdapter.diskUsageStats();
+
+        assertThat(diskUsageStats)
+                .hasSize(1)
+                .noneSatisfy(
+                        diskStats -> assertThat(diskStats.name()).isEqualTo("nodeWithMissingDiskStatistics")
+                )
+                .first()
+                .satisfies(
+                        nodeDescr -> {
+                            assertThat(nodeDescr.name()).isEqualTo("nodeWithCorrectInfo");
+                            assertThat(nodeDescr.ip()).isEqualTo("182.88.0.2");
+                            assertThat(nodeDescr.roles()).isEqualTo(NodeRole.parseSymbolString("dimr"));
+                            assertThat(nodeDescr.diskUsed().getBytes()).isEqualTo(SIUnitParser.parseBytesSizeValue("45gb").getBytes());
+                            assertThat(nodeDescr.diskTotal().getBytes()).isEqualTo(SIUnitParser.parseBytesSizeValue("411.5gb").getBytes());
+                            assertThat(nodeDescr.diskUsedPercent()).isEqualTo(10.95d);
+                        }
+                );
+
+    }
+}

--- a/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ClusterAdapterES7.java
+++ b/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/ClusterAdapterES7.java
@@ -104,7 +104,14 @@ public class ClusterAdapterES7 implements ClusterAdapter {
     }
 
     private List<NodeResponse> nodes() {
-        return catApi.nodes();
+        final List<NodeResponse> allNodes = catApi.nodes();
+        final List<NodeResponse> nodesWithDiskStatistics = allNodes.stream().filter(NodeResponse::hasDiskStatistics).collect(Collectors.toList());
+        if (allNodes.size() != nodesWithDiskStatistics.size()) {
+            final List<NodeResponse> nodesWithMissingDiskStatistics = allNodes.stream().filter(nr -> !nr.hasDiskStatistics()).collect(Collectors.toList());
+            LOG.info("_cat/nodes API has returned " + nodesWithMissingDiskStatistics.size() + " nodes without disk statistics:");
+            nodesWithMissingDiskStatistics.forEach(node -> LOG.info(node.toString()));
+        }
+        return nodesWithDiskStatistics;
     }
 
     @Override

--- a/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/cat/NodeResponse.java
+++ b/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/cat/NodeResponse.java
@@ -18,6 +18,7 @@ package org.graylog.storage.elasticsearch7.cat;
 
 import com.fasterxml.jackson.annotation.JsonAutoDetect;
 import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.auto.value.AutoValue;
 
@@ -37,12 +38,16 @@ public abstract class NodeResponse {
 
     public abstract String ip();
 
+    @Nullable
     public abstract String diskUsed();
 
+    @Nullable
     public abstract String diskTotal();
 
+    @Nullable
     public abstract Double diskUsedPercent();
 
+    @Nullable
     public abstract Long fileDescriptorMax();
 
     @JsonCreator
@@ -51,10 +56,10 @@ public abstract class NodeResponse {
                                       @JsonProperty("role") String role,
                                       @JsonProperty("host") @Nullable String host,
                                       @JsonProperty("ip") String ip,
-                                      @JsonProperty("diskUsed") String diskUsed,
-                                      @JsonProperty("diskTotal") String diskTotal,
-                                      @JsonProperty("diskUsedPercent") Double diskUsedPercent,
-                                      @JsonProperty("fileDescriptorMax") Long fileDescriptorMax) {
+                                      @JsonProperty("diskUsed") @Nullable String diskUsed,
+                                      @JsonProperty("diskTotal") @Nullable String diskTotal,
+                                      @JsonProperty("diskUsedPercent") @Nullable Double diskUsedPercent,
+                                      @JsonProperty("fileDescriptorMax") @Nullable Long fileDescriptorMax) {
         return new AutoValue_NodeResponse(
                 id,
                 name,
@@ -66,5 +71,13 @@ public abstract class NodeResponse {
                 diskUsedPercent,
                 fileDescriptorMax
         );
+    }
+
+    @JsonIgnore
+    public boolean hasDiskStatistics() {
+        return diskUsed() != null &&
+                diskTotal() != null &&
+                diskUsedPercent() != null &&
+                fileDescriptorMax() != null;
     }
 }

--- a/graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/ClusterAdapterES7Test.java
+++ b/graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/ClusterAdapterES7Test.java
@@ -21,18 +21,26 @@ import com.github.joschi.jadconfig.util.Duration;
 import com.google.common.io.Resources;
 import org.graylog.shaded.elasticsearch7.org.elasticsearch.ElasticsearchException;
 import org.graylog.storage.elasticsearch7.cat.CatApi;
+import org.graylog.storage.elasticsearch7.cat.NodeResponse;
+import org.graylog2.indexer.cluster.health.NodeDiskUsageStats;
+import org.graylog2.indexer.cluster.health.NodeFileDescriptorStats;
+import org.graylog2.indexer.cluster.health.NodeRole;
+import org.graylog2.indexer.cluster.health.SIUnitParser;
 import org.graylog2.indexer.indices.HealthStatus;
 import org.graylog2.shared.bindings.providers.ObjectMapperProvider;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableList;
 
 import java.io.IOException;
 import java.util.Collections;
 import java.util.Optional;
+import java.util.Set;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
@@ -45,6 +53,25 @@ class ClusterAdapterES7Test {
     private ObjectMapper objectMapper = new ObjectMapperProvider().get();
 
     private ClusterAdapterES7 clusterAdapter;
+
+    private final static NodeResponse NODE_WITH_CORRECT_INFO = NodeResponse.create("nodeWithCorrectInfo",
+            "nodeWithCorrectInfo",
+            "dimr",
+            null,
+            "182.88.0.2",
+            "45gb",
+            "411.5gb",
+            10.95d,
+            1048576L);
+    private final static NodeResponse NODE_WITH_MISSING_DISK_STATISTICS = NodeResponse.create("nodeWithMissingDiskStatistics",
+            "nodeWithMissingDiskStatistics",
+            "dimr",
+            null,
+            "182.88.0.1",
+            null,
+            null,
+            null,
+            null);
 
     @BeforeEach
     void setUp() {
@@ -82,6 +109,51 @@ class ClusterAdapterES7Test {
         when(client.execute(any())).thenThrow(new ElasticsearchException("Exception"));
         final Optional<HealthStatus> healthStatus = clusterAdapter.health(Collections.singletonList("foo_index"));
         assertThat(healthStatus).isEmpty();
+    }
+
+    @Test
+    void testFileDescriptorStats() {
+        doReturn(ImmutableList.of(NODE_WITH_CORRECT_INFO, NODE_WITH_MISSING_DISK_STATISTICS)).when(catApi).nodes();
+        final Set<NodeFileDescriptorStats> nodeFileDescriptorStats = clusterAdapter.fileDescriptorStats();
+
+        assertThat(nodeFileDescriptorStats)
+                .hasSize(1)
+                .noneSatisfy(
+                        nodeDescr -> assertThat(nodeDescr.name()).isEqualTo("nodeWithMissingDiskStatistics")
+                )
+                .first()
+                .satisfies(
+                        nodeDescr -> {
+                            assertThat(nodeDescr.name()).isEqualTo("nodeWithCorrectInfo");
+                            assertThat(nodeDescr.ip()).isEqualTo("182.88.0.2");
+                            assertThat(nodeDescr.fileDescriptorMax()).isPresent();
+                            assertThat(nodeDescr.fileDescriptorMax().get()).isEqualTo(1048576L);
+                        }
+                );
+    }
+
+    @Test
+    void testDiskUsageStats() {
+        doReturn(ImmutableList.of(NODE_WITH_CORRECT_INFO, NODE_WITH_MISSING_DISK_STATISTICS)).when(catApi).nodes();
+        final Set<NodeDiskUsageStats> diskUsageStats = clusterAdapter.diskUsageStats();
+
+        assertThat(diskUsageStats)
+                .hasSize(1)
+                .noneSatisfy(
+                        diskStats -> assertThat(diskStats.name()).isEqualTo("nodeWithMissingDiskStatistics")
+                )
+                .first()
+                .satisfies(
+                        nodeDescr -> {
+                            assertThat(nodeDescr.name()).isEqualTo("nodeWithCorrectInfo");
+                            assertThat(nodeDescr.ip()).isEqualTo("182.88.0.2");
+                            assertThat(nodeDescr.roles()).isEqualTo(NodeRole.parseSymbolString("dimr"));
+                            assertThat(nodeDescr.diskUsed().getBytes()).isEqualTo(SIUnitParser.parseBytesSizeValue("45gb").getBytes());
+                            assertThat(nodeDescr.diskTotal().getBytes()).isEqualTo(SIUnitParser.parseBytesSizeValue("411.5gb").getBytes());
+                            assertThat(nodeDescr.diskUsedPercent()).isEqualTo(10.95d);
+                        }
+                );
+
     }
 
     private void mockNodesResponse() throws IOException {

--- a/graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/cat/CatApiTest.java
+++ b/graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/cat/CatApiTest.java
@@ -1,3 +1,19 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
 package org.graylog.storage.elasticsearch7.cat;
 
 import com.fasterxml.jackson.core.type.TypeReference;

--- a/graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/cat/CatApiTest.java
+++ b/graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/cat/CatApiTest.java
@@ -1,0 +1,70 @@
+package org.graylog.storage.elasticsearch7.cat;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.graylog.storage.elasticsearch7.ElasticsearchClient;
+import org.graylog2.shared.bindings.providers.ObjectMapperProvider;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class CatApiTest {
+
+    private final static String SAMPLE_CAT_NODES_RESPONSE = "[" +
+            "{" +
+            "\"id\":\"nodeWithCorrectInfo\"," +
+            "\"name\":\"nodeWithCorrectInfo\"," +
+            "\"role\":\"dimr\"," +
+            "\"ip\":\"182.88.0.2\"," +
+            "\"fileDescriptorMax\":\"1048576\"," +
+            "\"diskUsed\":\"45gb\"," +
+            "\"diskTotal\":\"411.5gb\"," +
+            "\"diskUsedPercent\":\"10.95\"" +
+            "}," +
+            "{" +
+            "\"id\":\"nodeWithMissingDiskStatistics\"," +
+            "\"name\":\"nodeWithMissingDiskStatistics\"," +
+            "\"role\":\"dimr\"," +
+            "\"ip\":\"182.88.0.1\"" +
+            "}" +
+            "]";
+
+    @Test
+    void testNodesMethodParsesAndReturnsEvenNodesThatMissDiskUsageInfo() throws Exception {
+        final ObjectMapper objectMapper = new ObjectMapperProvider().get();
+        final ElasticsearchClient client = mock(ElasticsearchClient.class);
+        final CatApi toTest = new CatApi(objectMapper, client);
+
+        when(client.execute(any(), anyString()))
+                .thenReturn(objectMapper.readValue(SAMPLE_CAT_NODES_RESPONSE, new TypeReference<List<NodeResponse>>() {}));
+
+        final List<NodeResponse> nodes = toTest.nodes();
+
+        assertThat(nodes)
+                .hasSize(2)
+                .contains(NodeResponse.create("nodeWithCorrectInfo",
+                        "nodeWithCorrectInfo",
+                        "dimr",
+                        null,
+                        "182.88.0.2",
+                        "45gb",
+                        "411.5gb",
+                        10.95d,
+                        1048576L))
+                .contains(NodeResponse.create("nodeWithMissingDiskStatistics",
+                        "nodeWithMissingDiskStatistics",
+                        "dimr",
+                        null,
+                        "182.88.0.1",
+                        null,
+                        null,
+                        null,
+                        null));
+    }
+}

